src/halofit.jl

```diff
@@ -0,0 +1,211 @@
+    pk_NLs = transpose(reverse(pk_NLs, dims=1))
+    pk_NL = LinearInterpolation((k, z), pk_NLs)
+
+    PKnonlin(a, k, rsig, sigma2, neff, C, pk_NL)
+end 
+
+function gauss_norm_int_func(logk, pk, R)
+    k = exp.(logk)
+    k2 = k .^ 2
+    integ = @. pk*k*k2/2.0/pi^2*exp(-k2*R^2)
+    return integ
+end
+
+function onederiv_gauss_norm_int_func(logk, pk, R)
+    k = exp.(logk)
+    k2 = k .^ 2
+    integ = @. (-2.0*R*k2)*pk*k*k2/2.0/pi^2*exp(-k2*R^2)
+    return integ
+end
+
+function twoderiv_gauss_norm_int_func(logk, pk, R)
+    k = exp.(logk)
+    k2 = k .^ 2
+    R2 = R .^ 2
+    integ = @. (-2.0*k2 + 4*k2^2*R2)*pk*k*k2/2.0/pi^2*exp(-k2*R2)
+    return integ
+end
+
+# function whose root is \sigma^2{rsigma, a} = 1
+function rsigma_func(rsigma, pk, logk)
+    result = trapz(logk, gauss_norm_int_func(logk, pk, rsigma)) - 1.0
+    return result
+end
+
+function solve_for_rsigma(pk, logk)
+    rlow = 1e-2
+    rhigh = 1e2
+#     f1 = rsig -> rsigma_func(rsig, pk, logk)
+#     f2 = rsig -> trapz(logk, onederiv_gauss_norm_int_func(logk, pk, rsig))
+#     rsigma = find_zero((f1, f2), 5.5, Roots.Newton())
+    rsigma = find_zero(rsig -> rsigma_func(rsig, pk, logk), (rlow, rhigh))
+    return rsigma
+end
+
+function solve_for_rsigma(pk::Vector{D}, logk) where D<:ForwardDiff.Dual
+    pk0 = ForwardDiff.value.(pk)
+    rsig0 = solve_for_rsigma(pk0, logk)
+    ∂pk = ForwardDiff.gradient(pkl -> rsigma_func(rsig0, pkl, logk), pk0)
+    ∂rsig = ForwardDiff.derivative(rsig -> rsigma_func(rsig, pk0, logk), rsig0)
+    parts_in = zip(collect.(ForwardDiff.partials.(pk))...)
+    parts_out = map(x->dot(-∂pk./∂rsig, x), parts_in)
+    # (parts_out...,) from vector to (*, *, *)
+    p = ForwardDiff.Partials((parts_out...,))
+    return D(rsig0, p)
+end
+
+function get_rsigma(pk, logk)
+    rlow = 1e-2
+    rhigh = 1e2
+    # we have to bound the root, otherwise return -1
+    # we will fiil in any -1's in the calling routine
+    flow = rsigma_func(rlow, pk, logk)
+    fhigh = rsigma_func(rhigh, pk, logk)
+
+    if flow * fhigh > 0
+        return -1
+    end
+
+    rsigma = solve_for_rsigma(pk, logk)
+
+    return rsigma
+end
+
+function power_spectrum_nonlin(cosmo::Cosmology, PkL, k, a, rsig, sigma2, neff, C)
+
+    zs = @. 1.0/a - 1.0
+    weffa = -1.0
+    omegaMz = omega_x(cosmo, zs, "m")
+    omegaDEwz = omega_x(cosmo, zs,"l")
+
+    # not using these to match CLASS better - might be a bug in CLASS
+    # weffa = gsl_spline_eval(hf->weff, a, NULL);
+    # omegaMz = gsl_spline_eval(hf->omeff, a, NULL);
+    # omegaDEwz = gsl_spline_eval(hf->deeff, a, NULL);
+
+    ksigma = @. 1.0 / rsig
+    neff2 = @. neff * neff
+    neff3 = @. neff2 * neff
+    neff4 = @. neff3 * neff
+
+    delta2_norm = @. k*k*k/2.0/pi/pi
+
+    # compute the present day neutrino massive neutrino fraction
+    # uses all neutrinos even if they are moving fast
+#     om_nu = cosmo.sum_nu_masses / 93.14 / cosmo.h / cosmo.h
+#     fnu = om_nu / (cosmo.Ωm)
```

Comment:
What happened here?

Reply:
These were commented by Andrina. 
We could ask her.